### PR TITLE
Georgi Shopov HW02

### PR DESCRIPTION
**Pull request by georgiiishopovv**

### CODE REVIEW

File: Main.hs [hunk 1]
```diff
@@ -11,28 +11,52 @@ import qualified Data.List as L
...
+productOfDifference x = 
+    let z = zipWith (-) x (drop 1 x)
+    in product z
+
+isSorted :: [Int] -> Bool
+isSorted [] = True
+isSorted [_] = True 
+isSorted xs = 
+    let z = zipWith (<=) xs (drop 1 xs)
+    in all (==True) z
```
Comment (alex28sh) [inline Main.hs:33]: you can implement `isSorted` in 1 line

File: Main.hs [hunk 2]
```diff
@@ -11,28 +11,52 @@ import qualified Data.List as L
...
+isSorted [] = True
+isSorted [_] = True 
+isSorted xs = 
+    let z = zipWith (<=) xs (drop 1 xs)
+    in all (==True) z 
 
 countElement :: Int -> [Int] -> Int 
-countElement = undefined
-
+countElement _ [] = 0
```
Comment (alex28sh) [inline Main.hs:36]: don't need this case

File: Main.hs [hunk 3]
```diff
@@ -11,28 +11,52 @@ import qualified Data.List as L
...
+countElement x y =
+    let z = filter (==x) y
+    in length z
+ 
 dotProduct :: [Int] -> [Int] -> Int 
-dotProduct = undefined
+dotProduct [] [] = 0
```
Comment (alex28sh) [inline Main.hs:42]: don't need this case

File: Main.hs [hunk 4]
```diff
@@ -11,28 +11,52 @@ import qualified Data.List as L
...
+dotProduct x y =
+    let z = zipWith (*) x y
+    in sum z
 
 applyAll :: [a -> b] -> a -> [b]
-applyAll = undefined
+applyAll [] _ = []
+applyAll (x:xs) y = x y : applyAll (xs) y
```
Comment (alex28sh) [inline Main.hs:49]: use higher-order functions here

File: Main.hs [hunk 5]
```diff
@@ -11,28 +11,52 @@ import qualified Data.List as L
...
 interleave :: [a] -> [a] -> [a] 
-interleave = undefined
+interleave xs ys = interleave' (take n xs) (take n ys)
+    where n = min (length xs) (length ys)
+
+interleave' :: [a] -> [a] -> [a]
+interleave' [] [] = []
+interleave' _ [] = []
+interleave' [] _ = []
+interleave' (x:xs) (y:ys) = x : y : interleave' xs ys
```
Comment (alex28sh) [inline Main.hs:59]: you can implement it in 1 line using higher-order functions discussed in the lecture

**Review by alex28sh — CHANGES_REQUESTED** [review]
3/8

File: Main.hs [hunk 6]
```diff
@@ -11,28 +11,32 @@ import qualified Data.List as L
...
 productOfDifference :: [Int] -> Int
-productOfDifference = undefined
+productOfDifference [] = error "List too short!"
+productOfDifference [_] = error "List too short!"
+productOfDifference x = product $ zipWith (-) x (drop 1 x)
 
-isSorted :: [Int] -> Bool 
-isSorted xs = undefined
+isSorted :: [Int] -> Bool
+isSorted xs = all (==True) $ zipWith (<=) xs (drop 1 xs)
```
Comment (kajigor) [inline Main.hs:27]: `all (==True)` could also be written as `all id` or, what is better, `and`.

**Comment by kajigor** [comment]
ok

**Pull request closed without merging**